pkg/controller/control.go

```diff
@@ -185,25 +190,78 @@ func (c *Controller) addReplica(address string, snapshotRequired bool, mode type
 	return c.addReplicaNoLock(newBackend, address, snapshotRequired, mode)
 }
 
-// Snapshot will try to freeze the filesystem of the volume if possible
-// and will fallback to a system level sync in all other cases
-func (c *Controller) Snapshot(name string, labels map[string]string) (string, error) {
+// If shouldFreeze, Snapshot attempts to freeze the mounted filesystem on the volume's root partition. If it fails to
+// do so, Snapshot attempts to clean up and returns an error. If not shouldFreeze or if a mounted filesystem is not
+// detected on the volume's root partition, Snapshot does a best effort sync and takes a snapshot.
+func (c *Controller) Snapshot(name string, labels map[string]string, shouldFreeze bool) (string, error) {
+	if name == "" {
+		name = lhutils.UUID()
+	}
+
 	log := logrus.WithFields(logrus.Fields{"volume": c.VolumeName, "snapshot": name})
 	log.Info("Starting snapshot")
 
-	log.Info("Requesting system sync before snapshot")
-	if err := lhns.Sync(); err != nil {
-		// sync should never fail though, so it more like due to the nsenter
-		log.WithError(err).Errorf("WARNING: failed to sync continuing with snapshot for %v", name)
+	var endpoint string
+	c.RLock()
+	// Check now to avoid freezing or syncing unnecessarily. Also check again later as originally designed.
+	err := c.canDoSnapshot()
+	if c.frontend.FrontendName() == types.EngineFrontendBlockDev {
+		// It is meaningless to try to freeze filesystems for a tgt-iscsi endpoint.
+		endpoint = c.Endpoint()
+	}
+	c.RUnlock()
+	if err != nil {
+		return "", err
 	}
 
-	c.Lock()
-	defer c.Unlock()
+	var mounted, frozen bool
+	freezePoint := filepath.Join(freezePointDirectory, filepath.Base(endpoint))
+	mounter := mount.New("")
+	exec := lhexec.NewExecutor()
+	defer func() {
+		if frozen {
+			util.AttemptUnfreezeFilesystem(freezePoint, exec, true, log)
+		}
+		if mounted {
+			attemptUnmountFilesystem(freezePoint, mounter, log)
+		}
+	}()
 
-	if name == "" {
-		name = lhutils.UUID()
+	// We create a bind mount of one of the discovered mount points (it does not matter which one) inside the container
+	// mount namespace (assuming longhorn-engine is running in a container). The mount point cannot be unmounted except
+	// by us, so we can be confident it is safe to freeze. This approach is preferable to locking some source mount
+	// point with an open file descriptor because it cannot be circumvented by a lazy unmount at the source.
+	if shouldFreeze && endpoint != "" {
+		if !c.snapshotFreezeLock.TryLock() {
```

Comment:
We discussed this offline. While there are some pros and cons to both behaviors, for now we will leave it as a `TryLock`. It is mainly because `Lock` may be stuck forever in some unknown scenario, but the `Snapshot` client will give up waiting. In that case, the user will not know why the snapshot failed and the goroutine will exist indefinitely. It may be better to just tell the user the snapshot failed immediately and allow them to retry. 

FWIW, the argument for `Lock` is quite valid as well. We can almost certainly get the lock if we are patient, so we don't really need to return failure. If we notice that the chosen behavior is causing problems, we can reevaluate.